### PR TITLE
NAVAND-538: add MapboxNavigation#registerOffboardRoutesObserver

### PR DESCRIPTION
**Pull request by dzinad**

### CODE REVIEW

File: libnavigation-core/api/current.txt [hunk 1]
```diff
@@ -782,6 +791,10 @@ package com.mapbox.navigation.core.routealternatives {
     method public void onRouteAlternativesAborted(String message);
   }
 
+  public final class Test {
```
Comment (RingerJK) [inline libnavigation-core/api/current.txt:794]: 🤔

Comment (dzinad) [reply libnavigation-core/api/current.txt]: It happens...

**Comment by LukasPaczos** [comment]
I'm not sure whether adding a new, independent observer is the way to go. The feature is tightly tied to the alternatives observer, could we try to incorporate the information into the existing system?

I understand the `NavigationRouteAlternativesObserver` is the limiting factor here but what if we extended `NavigationRoute` to expose the `RouterOrigin`? Then it would be matter of a simple if-check in the `NavigationRouteAlternativesObserver` to see if the route is offboard or not, without having to introduce a whole new observer.

**Comment by dzinad** [comment]
> Then it would be matter of a simple if-check in the NavigationRouteAlternativesObserver to see if the route is offboard or not, without having to introduce a whole new observer.

It's already available:
```
     * Additionally, you can use the alternatives observer to switch to an offboard route if the current route was built onboard
     * which can be used to always prefer an offboard-generated route over onboard-generated one.
     * This is a good practice because offboard-generated routes take live road conditions into account,
     * have more precise ETAs, and can also be refreshed as the user drives and conditions change. Example usage:
     * ```kotlin
     * private val alternativesObserver = object : NavigationRouteAlternativesObserver {
     *     override fun onRouteAlternatives(
     *         routeProgress: RouteProgress,
     *         alternatives: List<NavigationRoute>,
     *         routerOrigin: RouterOrigin
     *     ) {
     *         val primaryRoute = routeProgress.navigationRoute
     *         val isPrimaryRouteOffboard = primaryRoute.origin == RouterOrigin.Offboard
     *         val offboardAlternatives = alternatives.filter { it.origin == RouterOrigin.Offboard }
     *
     *         when {
     *             isPrimaryRouteOffboard -> {
     *                 // if the current route is offboard, keep it
     *                 // but consider accepting additional offboard alternatives only and ignore onboard ones
     *                 val updatedRoutes = mutableListOf<NavigationRoute>()
     *                 updatedRoutes.add(primaryRoute)
     *                 updatedRoutes.addAll(offboardAlternatives)
     *                 mapboxNavigation.setNavigationRoutes(updatedRoutes)
     *             }
     *             isPrimaryRouteOffboard.not() && offboardAlternatives.isNotEmpty() -> {
     *                 // if the current route is onboard, and there's an offboard route available
     *                 // consider notifying the user that a more accurate route is available and whether they'd want to switch
     *                 // or force the switch like presented
     *                 mapboxNavigation.setNavigationRoutes(offboardAlternatives)
     *             }
     *             else -> {
     *                 // in other cases, when current route is onboard and there are no offboard alternatives,
     *                 // just append the new alternatives
     *                 val updatedRoutes = mutableListOf<NavigationRoute>()
     *                 updatedRoutes.add(primaryRoute)
     *                 updatedRoutes.addAll(alternatives)
     *                 mapboxNavigation.setNavigationRoutes(updatedRoutes)
     *             }
     *         }
     *     }
     *
     *     override fun onRouteAlternativesError(error: RouteAlternativesError) {
     *         // error handling
     *     }
     * }
```
But it's too much to do for the user. The goal of this PR is to provide a way to do it much simpler.

I don't see anything wrong with the new observer. Yes, it's connected to alternatives but also routeProgress is connected to routes but it doesn't mean we should have only 1 observer.

**Comment by LukasPaczos** [comment]
Adding a new observer would make it simper to directly access the data (a new offboard alternative route), however, it makes it more complicated to actually integrate that all together.

In an application setup I'd imagine developer observing both `NavigationRouteAlternativesObserver` and the `OffboardRoutesObserver`, from both of these observers they would attempt to update navigation routes. These updates would effectively compete with each other, especially with `setNavigationRoutes` being async, and I see a lot of races that developers would have to resolve themselves to keep the state of routes predictable. Having a single observer makes the integration process simpler. If we think that the suggested implementation needed to understand which are the new offboard alternatives routes is too complex now (which I agree with), let's add utility functions to encapsulate that logic and make it more accessible.

**Comment by dzinad** [comment]
> In an application setup I'd imagine developer observing both NavigationRouteAlternativesObserver and the OffboardRoutesObserver, from both of these observers they would attempt to update navigation routes. These updates would effectively compete with each other, especially with setNavigationRoutes being async, and I see a lot of races that developers would have to resolve themselves to keep the state of routes predictable. 

Ooooh, I see... I though we did not receive `onAlternativesChanged` if we received `onOnlinePrimaryRouteAvailable`... Because I didn't see the debug log (probably because of incorrect log level).
The you're right. Actually, it's a weird NN API. Receiving 2 callbacks for the same update is weird. 
@DenisPryt Is it safe to assume that in case we receive `onOnlinePrimaryRouteAvailable` in `RouteAlternativesObserver` we always receive `onRouteAlternativesChanged` as well?

**Comment by dzinad** [comment]
Closing this one as the logic will be different.

**Pull request closed without merging**